### PR TITLE
[for 4.6.0] Fix concurrent problem in ProducerManager.getAvaliableCha…

### PR DESCRIPTION
**Pull request by areyouok**
Fix two problem:
* If client channel is in writing and there is no more channel, the tx check message drops.
* ProducerManager.getAvaliableChannel/findChannel is not thread safe.

### CODE REVIEW

File: broker/src/main/java/org/apache/rocketmq/broker/client/ProducerManager.java [hunk 1]
```diff
@@ -36,205 +32,145 @@
...
+            if (channelTable.isEmpty()) {
+                this.groupChannelTable.remove(group);
+                log.info("unregister a producer group[{}] from groupChannelTable", group);
             }
-        } catch (InterruptedException e) {
-            log.error("", e);
         }
     }
 
     public Channel getAvaliableChannel(String groupId) {
```
[the remark this answers is not in the document]
Comment (areyouok) [reply broker/src/main/java/org/apache/rocketmq/broker/client/ProducerManager.java]: This method is public, change name may cause incompatibility

File: broker/src/main/java/org/apache/rocketmq/broker/client/ProducerManager.java [hunk 2]
```diff
@@ -36,205 +32,145 @@
...
-        HashMap<Channel, ClientChannelInfo> channelClientChannelInfoHashMap = groupChannelTable.get(groupId);
+        if (groupId == null) {
+            return null;
+        }
         List<Channel> channelList = new ArrayList<Channel>();
+        ConcurrentHashMap<Channel, ClientChannelInfo> channelClientChannelInfoHashMap = groupChannelTable.get(groupId);
         if (channelClientChannelInfoHashMap != null) {
             for (Channel channel : channelClientChannelInfoHashMap.keySet()) {
                 channelList.add(channel);
             }
```
[the remark this answers is not in the document]
Comment (areyouok) [reply broker/src/main/java/org/apache/rocketmq/broker/client/ProducerManager.java]: These are exist code and does not affect the target of this PR.

File: broker/src/main/java/org/apache/rocketmq/broker/client/ProducerManager.java [hunk 3]
```diff
@@ -36,205 +32,145 @@
...
+
+        int index = positiveAtomicCounter.incrementAndGet() % size;
+        Channel channel = channelList.get(index);
+        int count = 0;
+        boolean isOk = channel.isActive() && channel.isWritable();
+        while (count++ < GET_AVALIABLE_CHANNEL_RETRY_COUNT) {
+            if (isOk) {
+                return channel;
+            }
+            if (channel.isActive()) {
```
Comment (duhenglucky) [inline broker/src/main/java/org/apache/rocketmq/broker/client/ProducerManager.java]: This change will cause this method returns an active but not writable channel.

File: broker/src/main/java/org/apache/rocketmq/broker/client/ProducerManager.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vongosling) [reply broker/src/main/java/org/apache/rocketmq/broker/client/ProducerManager.java]: @areyouok there is no harm for your pr in here. Keeping simple code, write less code is more attractive.

**Pull request merged**